### PR TITLE
Prettier

### PR DESCRIPTION
**Pull request by kennylavender**
Considering adding prettier to the project again.

## Why

- Manually formatting is time consuming
- Thinking about the formatting is added overhead
- Consistency

## Previous issues

Previously I was using the prettier editor plugin to format the code. That resulted in each developer having their own version of prettier. So on occasion a developer may save a file with a small change and then prettier would adjust irrelevant lines of code to the change because of the prettier version differences.

## Solution

I think using the CLI is a better solution than the editor plugins. I have not figured out a good automatic way of running it yet. Might just need to do it manually.

### pre-commit
I was considering using a pre-commit hook to run prettier. The problem with a pre-commit hook though is that it requires `git add .` be ran after prettier to save the prettier changes to files. Using `git add .` might unintentionally add files that the developer changed and has intentionally left out of the commit, such as test imports file being modified to only seeing specific unit tests.

### CODE REVIEW

File: src/get-is-enabled.js [hunk 1]
```diff
@@ -4,13 +4,15 @@ import { find, curry } from 'ramda';
 const matchName = s => x => x.name === s;
 
 // enabled = x: Feature => boolean
-const enabled = x => x && x.enabled ? true : false;
+const enabled = x => (x && x.enabled ? true : false);
 
 // hasDependencies = x: Feature => boolean
-const hasDependencies = x => x.dependencies && x.dependencies.length > 0 ? true : false;
+const hasDependencies = x =>
+  x.dependencies && x.dependencies.length > 0 ? true : false;
```
Comment (ericelliott) [inline src/get-is-enabled.js]: You can delete  `? true : false` here. Comparators always evaluate to `true` or `false` by default.

**Comment by ericelliott** [comment]
Your comment mentioned some confusion about how to handle prettier. Can you walk me through the conclusions you came to?

**Comment by kennylavender** [comment]
Here are the 2 ways it can be automated from the project without relying on a editor plugin.

#### Watcher

We can use `onchange` to format when a file is saved. This would be nice.

```json
"scripts": {
    "prettier-watch": "onchange '**/*.js' -- prettier --write {{changed}}"
  },
```

It might require running another watch task from another terminal tab. But I might be able to combine all the watch tasks into a single command using something like [concurrently](https://www.npmjs.com/package/concurrently).


#### Git pre-hook

This is would be good, except it would add files to that commit that may have intentionally been left out of the commit (ex: a modified test import file while working on a single module).

```json
{
  "scripts": {
    "prettier": "prettier \"*/**/*.js\" --ignore-path ./.prettierignore --write && git add . && git status"
  },
  "pre-commit": [
    "prettier"
  ]
}
```

Il try these methods out once I get a chance let me know if you have any insights.

**Comment by kennylavender** [comment]
I've added prettier to `pre-commit` I think I like this for now, we just need to be careful of it adding unintentional unstaged changes.

**Comment by kennylavender** [comment]
Been messing with `lint-staged` and `husky` from the [prettier](https://prettier.io/docs/en/precommit.html) instructions and seems it does not have the issue of adding unstaged files by accident

**Comment by ericelliott** [comment]
Will prettier in the IDEs automatically read the config from the project?

**Comment by kennylavender** [comment]
Yes they do, but when developers have different prettier versions installed they fight with each other because the code output can be slightly different. 

I realized this when @thoragio and I were working on projects together. 

Every time one of us saved a file that that the other last touched it would make unnecessary formatting changes, it got really annoying to the point I uninstalled the plugin.

**Comment by kennylavender** [comment]
I got husky and lint-staged working on my work project today seems to work really well and doesn't have the issue of accidentally adding non staged files. So I just switched this pr over to that.

You may need to delete your node_modules and install fresh if there are any issues.

**Comment by kennylavender** [comment]
@thoragio let me know if it works for you.

**Comment by kennylavender** [comment]
nvm. just realized lint-staged and husky won't run tests easily. will try something else.

**Comment by kennylavender** [comment]
Prettier just updated their docs, perfect timing

They added [pretty-quick](https://prettier.io/docs/en/precommit.html#option-2-pretty-quick-https-githubcom-azz-pretty-quick)

Solves the issue and allows us to keep using pre-commit

**Comment by ericelliott** [comment]
This sounds good. So, let me see if I understand this:

1. We DON'T need to install prettier in the editors 
2. pretty-quick will fix things up pre-commit
3. We are NOT running a prettier watch task?

**Comment by kennylavender** [comment]
Correct!

**Comment by ericelliott** [comment]
Let's give it a shot.

**Comment by thoragio** [comment]
@kennylavender I've been playing around with this in a test project and it works really well. Nice!

One thing I have noticed is that if `eslint` is run before `pretty-quick` you can end up with code that does not meet all of the linting rules being checked into the repo. The reason is that sometimes prettier will make a change that does not comply with the rules eslint is following, so when you next view the changed file in your editor you will see the linting warnings/errors.

One option would be to change the order of the pre-commit sequence to run `pretty-quick` first and then run `eslint`, so that you get a console warning before committing the code. The only option really at that point is to add `eslint-disable` comments around the offending code, but this at least gets rid off the lint warning when editing the file in your editor, and you are then aware of the inconsistency between the prettier config and your linting rules. And if you're lucky there is a way to either change your prettier config or your eslint rules so they are consistent again (but this is not always possible).

I bring this up because I ran into an indent rule issue between prettier and the eslint rules I am running (Eric's standard pack as seen in greenrhum for example). [The issue has been floating around for a while but still has not been addressed](https://github.com/eslint/eslint/issues/6606). My only option here was to `eslint-disable` the offending code.

**Comment by kennylavender** [comment]
cool glad its working well!

Which code is getting eslint warnings or errors?

I removed all eslint formatting rules on purpose. eslint formatting rules such as indentation and prettier do not work well together and is a constant pain point.

In my opinion, and also by recommendation by [prettier](https://prettier.io/docs/en/eslint.html#turn-off-eslint-s-formatting-rules), eslint shouldn't be used for formatting rules when prettier is involved because prettier is in control of the format completely so there is no reason to have eslint validate format.

eslint should only be used for other types code issues, unused and undefined vars etc.

**Comment by thoragio** [comment]
eslint was flagging up a violation of the `indent` rule and was expecting one indent less for this block of code:

```
      const env = serverRendered
        ?
          {
            db: {
              apiKey: process.env.GREENRUHM_FIREBASE_API_KEY,
              authDomain: process.env.GREENRUHM_FIREBASE_AUTH_DOMAIN,
              databaseURL: process.env.GREENRUHM_FIREBASE_DATABASE_URL,
              projectId: process.env.GREENRUHM_FIREBASE_PROJECT_ID,
              storageBucket: process.env.GREENRUHM_FIREBASE_STORAGE_BUCKET,
              messagingSenderId:
                process.env.GREENRUHM_FIREBASE_MESSAGE_SENDER_ID
            }
          }
        :
          {};
```

It wanted it formatted like this:

```
      const env = serverRendered
        ?
        {
          db: {
            apiKey: process.env.GREENRUHM_FIREBASE_API_KEY,
            authDomain: process.env.GREENRUHM_FIREBASE_AUTH_DOMAIN,
            databaseURL: process.env.GREENRUHM_FIREBASE_DATABASE_URL,
            projectId: process.env.GREENRUHM_FIREBASE_PROJECT_ID,
            storageBucket: process.env.GREENRUHM_FIREBASE_STORAGE_BUCKET,
            messagingSenderId:
              process.env.GREENRUHM_FIREBASE_MESSAGE_SENDER_ID
          }
        }
        :
        {};
```

Personally, I agree with the prettier's recommendation (the first block above), so I used `/* eslint-disable indent */`/`/* eslint-enable indent */`

**Comment by kennylavender** [comment]
Yup, you guys should remove the eslint formatting rules from that project! 👍

**Comment by thoragio** [comment]
Definitely worth doing if/when we decide to integrate prettier into it!

**Pull request merged**